vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/treegrid/demo/TreeGridDemo.java

```diff
@@ -146,4 +149,78 @@ public boolean hasChildren(Account item) {
 
         addCard("TreeGrid with lazy loading", grid);
     }
+
+    // Drag and drop
+    private Department draggedItem;
+
+    private void createDragAndDrop() {
+        DepartmentData departmentData = new DepartmentData();
+        // begin-source-example
+        // source-example-heading: Drag and drop
+        TreeGrid<Department> treeGrid = new TreeGrid<>();
+
+        treeGrid.setDataProvider(
+                new AbstractHierarchicalDataProvider<Department, Object>() {
+                    @Override
+                    public int getChildCount(
+                            HierarchicalQuery<Department, Object> query) {
+                        return departmentData.getChildCount(query.getParent());
+                    }
+
+                    @Override
+                    public Stream<Department> fetchChildren(
+                            HierarchicalQuery<Department, Object> query) {
+                        return departmentData
+                                .getChildDepartments(query.getParent())
+                                .stream();
+                    }
+
+                    @Override
+                    public boolean hasChildren(Department item) {
+                        return departmentData.hasChildren(item);
+                    }
+
+                    @Override
+                    public boolean isInMemory() {
+                        return false;
+                    }
+                });
+
+        treeGrid.addHierarchyColumn(Department::getName)
+                .setHeader("Department Name");
+        treeGrid.addColumn(Department::getManager).setHeader("Manager");
+        treeGrid.setRowsDraggable(true);
+        treeGrid.setDropMode(GridDropMode.ON_TOP);
+
+        treeGrid.addDragStartListener(
+                event -> draggedItem = event.getDraggedItems().get(0));
+
+        treeGrid.addDragEndListener(event -> draggedItem = null);
+
+        treeGrid.addDropListener(event -> {
+            Department department = event.getDropTargetItem().get();
```

Comment:
the variable needs a better name, it wasn't immediately clear what this code was doing.

Is there a way of preventing this drop action completely?  with specifying allowed drop targets or something?

Reply:
Variable name changed.

I have no idea.